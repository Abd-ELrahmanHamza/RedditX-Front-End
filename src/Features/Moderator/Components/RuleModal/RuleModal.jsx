import React, { useState, useEffect } from "react";
import Form from "react-bootstrap/Form";
import Modal from "react-bootstrap/Modal";
import useFetchFunction from "Hooks/useFetchFunction";
import { useAuth } from "Features/Authentication/Contexts/Authentication";
import getSubreddit from "Features/Subreddit/Services/getSubreddit";
import { useParams } from "react-router-dom";
import createRule from "Features/Moderator/Services/createRule";
import editRule from "Features/Moderator/Services/editRule";
import {
  Characters,
  FormCheck,
  FormLabel,
  ModalContainer,
  FormText,
  ReportLabel,
  ReportDefault,
  Footer,
  AddRuleBtn,
  CancelBtn,
  DeleteBtn,
  UsedBefore,
} from "./RuleModal.styled";

const RuleModal = ({
  showModal,
  closeModal,
  setShowModal,
  showEditModal,
  setShowEditModal,
  ruleData,
}) => {
  // get name of community
  const { subredditId } = useParams();
  // const [show, setShow] = useState(false);
  const [rule, setRule] = useState(ruleData !== null ? ruleData.title : "");
  const [reasonRule, setReasonRule] = useState(
    ruleData !== null ? ruleData.description : ""
  );
  const [description, setDescription] = useState(
    ruleData !== null ? ruleData.reason : ""
  );
  const [ruleId, setRuleId] = useState(ruleData !== null ? ruleData._id : "");
  // state store type of community
  const [currentRadioValue, setCurrentRadioValue] = useState("posts-comments");
  // state for rule is used or not
  const [isUsedRule, setIsUsedRule] = useState(false);

  useEffect(() => {
    if (ruleData !== null) {
      setRule(ruleData.title);
      setDescription(ruleData.description);
      setReasonRule(ruleData.reason);
      setRuleId(ruleData._id);
    }
  }, [ruleData]);

  // authorization's user
  const auth = useAuth();
  // Fetch communities
  // Call useFetchFunction hook to handle states: loading, error, data
  // Loading: Boolean to tell if the request has been sent, or it's still loading
  // Error: Contains error message when the request is failed
  // Data: the response data
  const [rulesList, error, isLoading, fetchRules] = useFetchFunction();
  const [ruleResponse, errorResponse, isLoadingResponse, fetchData] =
    useFetchFunction();
  const [
    editRuleResponse,
    errorEditResponse,
    isLoadingEditResponse,
    fetcEdithData,
  ] = useFetchFunction();

  useEffect(() => {
    getSubreddit(fetchRules, `t5_${subredditId}`, auth);
  }, []);

  useEffect(() => {
    let flag = 0;
    if (
      !isLoading &&
      rulesList &&
      rulesList.things &&
      rulesList.things[0].communityRules
    ) {
      let result = rulesList.things[0].communityRules;
      result.map((el) => {
        if (el.title === rule && rule !== "" && ruleData.title !== el.title) {
          setIsUsedRule(true);
          flag = 1;
        }
      });
      if (!flag) {
        setIsUsedRule(false);
      }
    }
  }, [rule]);

  // create new rule
  const handleCreateRule = () => {
    if (!isUsedRule) {
      createRule(fetchData, auth, {
        srName: `t5_${subredditId}`,
        rule: {
          title: rule,
          description: description,
          reason: reasonRule,
        },
      });
      resetModal();
      closeModal();
    }
  };

  // handle edit rule
  const handleEditRule = () => {
    if (
      (ruleData.title !== rule && rule !== "") ||
      ruleData.description !== description ||
      ruleData.reason !== reasonRule
    ) {
      editRule(fetcEdithData, auth, {
        srName: `t5_${subredditId}`,
        rule: {
          title: rule,
          description: description,
          reason: reasonRule,
          id: ruleId,
        },
      });
      closeModal();
    }
  };

  // function that reset all values in modal
  const resetModal = () => {
    if (ruleData === null) {
      setRule("");
      setDescription("");
      setReasonRule("");
      setCurrentRadioValue("posts-comments");
      setShowEditModal(false);
    }
  };

  /**
   * function that controls length of string in modal's inputForm
   * @param event
   */
  const handlerRules = (event) => {
    setRule(event.target.value);
  };
  const handleReason = (event) => {
    setReasonRule(event.target.value);
  };
  const handleDescription = (event) => {
    setDescription(event.target.value);
  };
  return (
    <ModalContainer
      show={showModal}
      onHide={() => {
        resetModal();
        closeModal();
      }}
    >
      <Modal.Header closeButton>
        <Modal.Title>Edit rule</Modal.Title>
      </Modal.Header>
      <Modal.Body>
        <Form>
          <Form.Group className="mb-3">
            <Form.Label>Rule</Form.Label>
            <Form.Control
              as={"textarea"}
              placeholder='Rule displayed (e.g "No Photos")'
              autoFocus
              rows={2}
              value={rule}
              onChange={handlerRules}
              maxLength="100"
            />

            {isUsedRule && (
              <UsedBefore>
                You have another rule with this title. Please change.
              </UsedBefore>
            )}
            <Characters alarmValue={100 - rule.length}>
              {" "}
              {100 - rule.length} Characters remaining
            </Characters>
          </Form.Group>
          <FormCheck>
            <p>Applies to</p>
            <div>
              <FormLabel>
                <input
                  className="form-check-input"
                  type="radio"
                  id={"checkId-1"}
                  name="Radio"
                  value="posts-comments"
                  onChange={(e) => setCurrentRadioValue(e.target.value)}
                  defaultChecked={currentRadioValue === "posts-comments"}
                />
                <label htmlFor="checkId-1">
                  <FormText>Posts & comments</FormText>
                </label>
              </FormLabel>
              <FormLabel>
                <input
                  className="form-check-input"
                  type="radio"
                  id={"checkId-2"}
                  name="Radio"
                  value="posts-only"
                  onChange={(e) => setCurrentRadioValue(e.target.value)}
                  defaultChecked={currentRadioValue === "posts-only"}
                />
                <label htmlFor="checkId-2">
                  <FormText>Posts only</FormText>
                </label>
              </FormLabel>
              <FormLabel>
                <input
                  className="form-check-input"
                  type="radio"
                  id={"checkId-3"}
                  name="Radio"
                  value="comments-only"
                  onChange={(e) => setCurrentRadioValue(e.target.value)}
                  defaultChecked={currentRadioValue === "comments-only"}
                />
                <label htmlFor="checkId-3">
                  <FormText>Comments only</FormText>
                </label>
              </FormLabel>
            </div>
          </FormCheck>

          <Form.Group className="mb-3">
            <ReportLabel>Report reason</ReportLabel>
            <ReportDefault>Defaults to rule name if left blank</ReportDefault>
            <Form.Control
              as="textarea"
              rows={2}
              placeholder='Reason rule is broken (eg "This is a photo")'
              onChange={handleReason}
              maxLength="100"
              value={reasonRule}
            />
            <Characters alarmValue={100 - reasonRule.length}>
              {" "}
              {100 - reasonRule.length} Characters remaining
            </Characters>
          </Form.Group>
          <Form.Group className="mb-3">
            <ReportLabel>Full description</ReportLabel>
            <Form.Control
              as="textarea"
              rows={3}
              placeholder="Enter the full description of the rule"
              onChange={handleDescription}
              maxLength="500"
              value={description}
            />
            <Characters alarmValue={500 - description.length}>
              {" "}
              {500 - description.length} Characters remaining
            </Characters>
          </Form.Group>
        </Form>
      </Modal.Body>
      <Footer>
        {!showEditModal && (
          <AddRuleBtn
            onClick={handleCreateRule}
            addRule={rule.length > 0 && !isUsedRule}
          >
            Add new rule
          </AddRuleBtn>
        )}
        {showEditModal && (
          <AddRuleBtn
            onClick={handleEditRule}
            addRule={rule !== ruleData.title}
          >
            Save
          </AddRuleBtn>
        )}
        <CancelBtn
          onClick={() => {
            resetModal();
            setShowModal(false);
            setShowEditModal(false);
          }}
        >
          cancel
        </CancelBtn>
        <DeleteBtn show={showEditModal}>Delete</DeleteBtn>
      </Footer>
    </ModalContainer>
  );
};

export default RuleModal;
