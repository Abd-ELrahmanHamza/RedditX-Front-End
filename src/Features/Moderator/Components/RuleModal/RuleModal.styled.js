import styled from "styled-components";
import Modal from "react-bootstrap/Modal";
import Button from "react-bootstrap/Button";
import { ModalFooter } from "react-bootstrap";

export const ModalContainer = styled(Modal)`
  .modal-content {
    max-width: 538px !important;
    min-width: 410px !important;
  }

  .modal-header {
    border-bottom: 1px solid ${({ theme }) => theme.lineColor.primary};
    border-top-right-radius: 4px;
    border-top-left-radius: 4px;
    padding: 16px;

    .modal-title {
      font-size: 16px;
      font-weight: 500;
      line-height: 20px;
    }
    .btn-close:focus {
      box-shadow: none !important;
    }
  }
  .modal-body {
    .form-control {
      :focus {
        box-shadow: none;
      }
    }
  }
`;
export const Characters = styled.p`
  color: ${({ alarmValue }) =>
    alarmValue > 0 ? "#7c7c7c" : "#ea0027"} !important;
  padding-top: 5px;
  font-weight: 400;
  font-size: 12px;
  line-height: 16px;
`;
export const FormCheck = styled.div`
  display: block;
  margin-bottom: 16px;
  p {
    font-family: Noto Sans, Arial, sans-serif;
    font-size: 14px;
    font-weight: 400;
    line-height: 18px;
    padding-top: 8px;
  }
`;
export const FormLabel = styled.div`
  display: flex;
  margin-bottom: 16px;

  .form-check-input {
    width: 1em;
    height: 1em;

    :checked {
      background-color: #0079d3;
    }

    :focus {
      box-shadow: none;
    }
  }
`;
export const FormText = styled.div`
  display: inline;
  font-family: Noto Sans, Arial, sans-serif;
  font-weight: 400;
  color: ${({ theme }) => theme.categoryColor.text};
  font-size: 14px;
  line-height: 21px;
  margin-left: 6px;
`;
export const ReportLabel = styled.p`
  font-family: Noto Sans, Arial, sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 18px;
  padding-top: 8px;
`;
export const ReportDefault = styled.div`
  padding: 4px 0 8px;
  color: ${({ theme }) => theme.color.muted};
  font-size: 12px;
  font-weight: 400;
  line-height: 16px;
`;
export const ButtonContainer = styled(Button)`
  font-family: Noto Sans, Arial, sans-serif;
  font-size: 14px;
  font-weight: 700;
  letter-spacing: unset;
  line-height: 17px;
  text-transform: unset;
  min-height: 32px;
  min-width: 32px;
  padding: 4px 16px;
  cursor: not-allowed;
  filter: grayscale(1);
  color: ${({ theme }) => theme.borderColor.primary};
  border: 1px solid ${({ theme }) => theme.btnColor.primary};
  margin-left: 16px;
  position: relative;
  align-items: center;
  border-radius: 9999px;
  box-sizing: border-box;
  display: flex;
  justify-content: center;
  text-align: center;
  width: auto;
  //margin-right:${({ deleteBtn }) => (deleteBtn ? "auto" : "none")};
`;
export const Footer = styled(ModalFooter)`
  background-color: ${({ theme }) => theme.lineColor.primary};
  border-top: 1px solid ${({ theme }) => theme.lineColor.primary};
  flex-direction: row-reverse;
  justify-content: flex-start;
  display: flex;
  padding: 16px;
`;
