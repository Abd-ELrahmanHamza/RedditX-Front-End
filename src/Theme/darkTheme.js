/**
 * dark theme colors
 *
 * @type {{id:string}}
 */
const darkTheme = {
  // The id of the theme
  id: "dark",
  // The canvas color is the background color of the website
  canvas: "#030303",
  // The colors of the text on the website
  color: {
    // The primary color of the text on the website
    primary: "#d7dadc",
    // The color of the muted text on the website
    muted: "#7c7c7c",
    // The color of the links on the website
    link: "#0079d3",
  },
  // The colors of the backgrounds in the website
  background: {
    primary: "#1a1a1b",
  },
  // The colors of the lines in the website
  lineColor: {
    primary: "#343536",
  },
};

export default darkTheme;
