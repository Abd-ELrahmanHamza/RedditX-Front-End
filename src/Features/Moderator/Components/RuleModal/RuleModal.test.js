import React from "react";
import { BrowserRouter, Routes } from "react-router-dom/dist";
import darkTheme from "Theme/darkTheme";
import { ThemeProvider } from "styled-components";
import { screen, render, fireEvent, waitFor } from "@testing-library/react";
// Import contexts
import {
  AuthProvider,
  useAuth,
} from "Features/Authentication/Contexts/Authentication";

// import Components
import RuleModal from "./RuleModal";

jest.mock("Features/Authentication/Contexts/Authentication", () => ({
  __esModule: true, // this property makes it work
  useAuth: function () {
    return {
      getToken: jest.fn().mockReturnValue("token"),
    };
  },
  AuthProvider: ({ children }) => {
    return (
      <mock-authprovider data-testid="auth-provider">
        {children}
      </mock-authprovider>
    );
  },
}));
const ruleData = [
  {
    srName: "t5_mangaa",
    title: "Test",
    description: "management",
    reason: "test",
  },
];

describe("Rule Modal Component", () => {
  test("Rule Modal should be appeard", async () => {
    render(
      <BrowserRouter>
        <AuthProvider>
          <ThemeProvider theme={darkTheme}>
            <RuleModal ruleData={ruleData} />
          </ThemeProvider>
        </AuthProvider>
      </BrowserRouter>
    );
    const test = screen.getByText("Report reason");
    expect(test).toBeInTheDocument();
  });
});
