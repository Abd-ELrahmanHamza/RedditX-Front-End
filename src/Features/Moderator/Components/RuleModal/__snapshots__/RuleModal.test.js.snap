// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rule Modal Component this is a test for Rule Modal 1`] = `
<Router
  location={
    Object {
      "hash": "",
      "key": "default",
      "pathname": "/",
      "search": "",
      "state": null,
    }
  }
  navigationType="POP"
  navigator={
    Object {
      "action": "POP",
      "createHref": [Function],
      "go": [Function],
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "default",
        "pathname": "/",
        "search": "",
        "state": null,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Routes>
    <RuleModal />
  </Routes>
</Router>
`;
